text/2978-stack_based_vec.md

```diff
@@ -0,0 +1,364 @@
+
+```rust
+// `array_vec` has a pre-allocated memory of 2048 bits (32 * 64) that can store up
+// to 64 signed integers.
+let mut array_vec: ArrayVec<i32, 64> = ArrayVec::new();
+
+// Although reserved, there isn't anything explicitly stored yet
+assert_eq!(array_vec.len(), 0);
+
+// Initializes the first 32 bits with a simple '1' integer or
+// 00000000 00000000 00000000 00000001 bits
+array_vec.push(1);
+
+// Our vector memory is now split into a 32/2016 pair of initialized and
+// uninitialized memory respectively
+assert_eq!(array_vec.len(), 1);
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`ArrayVec` is a contiguous memory block where elements can be collected, therefore, a collection by definition and even though `core::collections` doesn't exist, it is the most natural module placement.
+
+To avoid length and conflicting conversations, the API will mimic most of the current `Vec` surface, which also means that all methods that depend on valid user input or valid internal capacity will panic at run-time when something goes wrong. For example, removing an element that is out of bounds.
+
+```rust
+// Please, bare in mind that these methods are simply suggestions. Discussions about the
+// API should probably take place elsewhere.
+
+pub struct ArrayVec<T, const N: usize> {
+    data: MaybeUninit<[T; N]>,
+    len: usize,
+}
+
+impl<T, const N: usize> ArrayVec<T, N> {
+    // Constructors
+
+    pub const fn new() -> Self;
+
+    // Infallible Methods
+
+    pub const fn as_mut_ptr(&mut self) -> *mut T;
+
+    pub const fn as_mut_slice(&mut self) -> &mut [T];
+
+    pub const fn as_ptr(&self) -> *const T;
+
+    pub const fn as_slice(&self) -> &[T];
+
+    pub const fn capacity(&self) -> usize;
+
+    pub fn clear(&mut self);
+
+    pub const fn is_empty(&self) -> bool;
+
+    pub const fn len(&self) -> usize;
+
+    pub fn retain<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut T) -> bool;
+
+    pub fn truncate(&mut self, len: usize);
+
+    // Methods that can panic at run-time
+
+    pub fn drain<R>(&mut self, range: R) -> Drain<'_, T, N>
+    where
+        R: RangeBounds<usize>;
+
+    pub fn extend_from_cloneable_slice<'a>(&mut self, other: &'a [T])
+    where
+        T: Clone;
+
+    pub fn extend_from_copyable_slice<'a>(&mut self, other: &'a [T])
+    where
+        T: Copy;
+
+    pub const fn insert(&mut self, idx: usize, element: T);
+
+    pub const fn push(&mut self, element: T);
+
+    pub const fn remove(&mut self, idx: usize) -> T;
+
+    pub fn splice<I, R>(&mut self, range: R, replace_with: I) -> Splice<'_, I::IntoIter, N>
+    where
+        I: IntoIterator<Item = T>,
+        R: RangeBounds<usize>;
+
+    pub fn split_off(&mut self, at: usize) -> Self;
+
+    pub fn swap_remove(&mut self, idx: usize) -> T;
+
+    // Verifiable methods
+    
+    pub const fn pop(&mut self) -> Option<T>;
+}
+```
```

Comment:
These APIs heavily mirror `Vec`. But with the introduction of const generic arrays there are other needs in the standard library that are not about providing a more lightweight Vec but instead a way to partially initialize arrays and safely drop them.

https://github.com/rust-lang/rust/issues/81615 is in search for a return type for cases where there may or may not have been enough elements to initialize an array of fixed size. `ArrayVec` could fill that role, but only when it has some additional const-generic functions such as safely unwrapping an array if and only if `len == N`

This gradual initialization could also be handled by `ArrayVec` (iirc  very similar code also exists in several other places):

https://github.com/rust-lang/rust/blob/186f7ae5b04d31d8ccd1746ac63cdf1ab4bc2354/library/core/src/array/mod.rs#L433-L463

Another feature with which this would intersect is SIMD. A partially filled `ArrayVec` is quite similar to SIMD value + mask.

Reply:
The conversion of a filled `ArrayVec` can be done like this:

```rust
impl<T, const N: usize> TryFrom<ArrayVec<T, N>> for [T; N] {
    type Error = ArrayVec<T, N>;
    fn try_from(other: ArrayVec<T, N>) -> Result<[T; N], ArrayVec<T, N>> {
        if other.len == N {
            let (array, _) = other.into_raw_parts();
            Ok(unsafe { transmute(array) })
        } else {
            Err(other)
        }
    }
}
```